src/hotspot/share/code/nmethod.cpp

```diff
@@ -474,13 +474,23 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
     if (exception_handler != -1) {
       offsets.set_value(CodeOffsets::Exceptions, exception_handler);
     }
-    nm = new (native_nmethod_size, CompLevel_none)
+
+    nm = new (native_nmethod_size, method->is_method_handle_intrinsic())
     nmethod(method(), compiler_none, native_nmethod_size,
             compile_id, &offsets,
             code_buffer, frame_size,
             basic_lock_owner_sp_offset,
             basic_lock_sp_offset,
             oop_maps);
+#ifdef ASSERT
+    // Method handle instrinsics may get allocated in NonNMethod space.
+    // GC may not look for Oops, there.
+    if (method->is_method_handle_intrinsic()) {
+      assert(oop_maps == nullptr, "expectation");
+      assert(!nm->contains_oops(), "no oops allowed");
+      assert(nm->metadata_size() == 0, "metadata usage not expected");
+    }
+#endif
```

Comment:
It might make sense to do this check for all nmethods that were allocated in NonNMethod.